### ISSUE TITLE
Some Issues in Reproduce

### ISSUE THREAD

**Issue by starforTJ** [ags3927/MERGE#3]
Aniruddha,

This is Taejin Jeong, who contacted you via email a few days ago.
As per your suggestion, I’m sharing a few questions here through GitHub Issues.

***

1. In the data preprocessing steps, the guidelines indicate converting JPG images to TIF format using the bash script provided by ST-Net. However, I encountered issues since this script relies on a module named "Histonet," which seems to be unavailable. I also noticed that this issue has been raised in GitHub Issues. Could you share how you personally handled the conversion from JPG to TIF?

2. Your paper compares two smoothing techniques, 8n and SPCS, ultimately choosing SPCS as the preferred method. However, in the provided code, the default setting appears to be spcs_to_8n. Could you clarify how spcs_to_8n was derived, and recommend which smoothing method should be used for reproduction?

3. Regarding other hyperparameters, is it advisable to proceed with the default values provided in your repository, or would you recommend any adjustments?

4. Could you explain what gnn_test_ext is used for? And during inference, which script should be used between gnn_block and gnn_block_whole_dataset?

***

My goal is to achieve a complete reproduction of Dr. Ganguly’s research, as I plan to reference it in my own study.
I also have a couple of follow-up questions:

5. As a follow-up to question 2, I now understand how the spcs_to_8n method works. However, I’d like to ask: during your inference, did you use the original spcs method or the spcs_to_8n method?

6. I noticed that the training pipeline runs experiments five times. However, even when using the provided seed values, the performance metrics (e.g., PCC, MSE, MAE) do not exactly match those reported in the CVPR paper. For example:
a. The PCC values do not match exactly.
b. The MSE and MAE values show larger differences than expected.
I suspect this may be due to averaging across folds or a different handling of the test results. Could you kindly explain the exact process you used to calculate the final performance values reported in the paper?

Thank you again for your support and for sharing the code. Looking forward to your clarification!

Best regards,
Taejin Jeong

**Comment by ags3927** [comment]
Dear Taejin,

Thank you so much for your questions. Here are the original responses to your first four questions.

1. In the preprocessing steps, you do not need to convert the images into TIF format. Even if it is a JPG image, the code extracts patches based on the tissue coordinates. The TIF conversion is an optional step, you can safely ignore it for the purposes of our problem setting. 
PS: TIF conversion is useful if you have multiple scaled versions of the WSI in a pyramid structure (5x, 10x, 20x, 40x, etc.). I provided the link in case anyone has access to multiple resolutions of images for any test data but does not have a full TIF image.

2. As mentioned in the [README description here](https://github.com/ags3927/MERGE/blob/master/README.md#the-counts_spcs_to_8n-directory), this directory contains the SPCS smoothed counts min-max scaled to the 8n smoothed counts. So the steps would be: i. Apply SPCS and 8n to the raw UMI counts to get two smoothed matrices, ii. Min-max scale the SPCS-smoothed matrix to fit the range of values in the 8n-smoothed matrix.

3. The hyperparameters in the config file are based on our experiments with the three datasets. If you are training on a new cohort or new dataset, I would suggest playing around with the values a little bit, specially the cluster counts. The cluster count of 5 is moderately dependent on the number of spots in the sample data. If you set a very low count, you will end up under-clustering the data and the various homogeneous tissue groups will not be separately clustered. If you set it too high though, it will over-cluster and that is not desirable either. Other parameters should be fine as they are. Maybe you can tweak the batch size or epochs but overall they should be fine.

4. The "gnn_test_ext" is a legacy mode from my old version of the code when I was still experimenting. I tried to write generalized versions of dataloader, cnn trainer and gnn trainer for any external dataset but realized that it would be better to just provide documentation for expected data format so that other people could convert to the necessary format. I have not set up a script for inference separately. If you are training on new data, you would have to first convert it into the expected data format (I can try to help you with that). And if you want to only run inference then you can change the train-test split code so that the entire list of slides is included in the val set and then use the pre trained checkpoints in gnn_test mode. Additionally, the "whole dataset" code is for people with adequate resources to load their entire dataset onto the GPU. But I realize that not everyone might have those resources, which is why I provided slide-by-slide training code by default. This has worse performance as pointed out in the GitHub issue#2 (which is now closed). You can use the whole dataset code by uncommenting it and then commenting out the next two function calls for the slide-by-slide mode.

<hr>

5. Could you explain what you mean by during inference? If you meant during validation tests, then yes, we used the min-max scaled values for validation for all datasets and all methods (all baselines and our own method).

6. You are most likely correct about the averaging issue. Each fold has five runs, and I find the best run in terms of lowest MSE from among the five runs for each fold. This gives me _k best runs_ for the k folds. Then I average these values for each metric. [This script](https://github.com/ags3927/MERGE/blob/master/scripts/find_best_results.ipynb) is the one for finding best run across all five runs. Then I just put all of those in a spreadsheet and find out the average values, standard deviation, etc.

Please reach out for any queries. I would love to help with any issue.

Regards,
Ani

**Comment by ags3927** [comment]
Btw, I am still a PhD candidate haha. So please, just call me Aniruddha or Ani. I am not Dr. Ganguly yet 😆

**Comment by starforTJ** [comment]
Ah, I see — thank you! 😊

I believe I may have phrased question 5 incorrectly.

I meant to ask about **the training step**, not inference.
In that case, should I understand that you also used **spcs_to_8n during training**?

Also, would it be possible for you to share the preprocessing code for spcs, 8n, and spcs_to_8n?

**Comment by ags3927** [comment]
Hi, yes you should also use the **spcs_to_8n** min-maxed counts for training.

As for the preprocessing:
1. I can provide you the 8n smoothing code. I wrote a python script for it.
2. For SPCS I used the original R code the authors provided in their [git repository](https://github.com/Usos/SPCS). You can simply convert your data to their format and then use it. I can give you a Jupyter notebook for inter-conversion between the expected data format of SPCS and what I use in MERGE.

I will upload these files in repository and update you here.

**Comment by starforTJ** [comment]
Thank you! Just to confirm — does spcs_to_8n mean that spcs is applied first, followed by 8n smoothing?

~~Also, while running gnn_train, I noticed that in some slides, there are certain genes that are not expressed in any of the spots — meaning the expression values for those genes are zero across the entire slide.
I understand that the nonzero filtering removes spots where all genes have zero expression, but it does not remove cases where a specific gene has zero expression across all spots in a slide.
In such cases, it becomes impossible to compute the Pearson correlation value.
How did you handle this situation?~~

I believe this issue might not have occurred because the training was done using the whole_dataset setting.

**Comment by ags3927** [comment]
Hi. I have added four new scripts.

1. [8n_smoothing.py](https://github.com/ags3927/MERGE/blob/master/scripts/8n_smoothing.py) is the python script for 8n smoothing. It expects the data to be in MERGE format.
2. [data_conv.ipynb](https://github.com/ags3927/MERGE/blob/master/scripts/data_conv.ipynb) is the notebook for converting MERGE format data to SPCS format data, and then the reverse too. Only run the necessary cells.
3. [smooth.ipynb](https://github.com/ags3927/MERGE/blob/master/scripts/smooth.ipynb) is the notebook for performing SPCS smoothing. But you will need to install R kernel inside your conda environment and the necessary R libraries to make it work. Just google how to install R kernel in conda and then follow the SPCS repository to install the necessary libraries. Or you can keep running until you hit an error for a missing package and then just install that one haha.
4. [min_max_scaler.ipynb](https://github.com/ags3927/MERGE/blob/master/scripts/min_max_scaler.ipynb) is the notebook for min-max scaling the SPCS smoothed data to the 8n range.

**Comment by ags3927** [comment]
To answer your earlier question, **spcs_to_8n** does not mean first SPCS and then 8n.

Let's say the original UMI counts or gene expression counts are called _counts_.
When you apply SPCS to _counts_, you get _counts_spcs_.
Similarly, when you apply 8n to _counts_, you get _counts_8n_.

Finally, you rescale the _counts_spcs_ values to match the data range in _counts_8n_, to obtain _counts_spcs_to_8n_.

This does not ruin the data distribution, but rescales the values so that the MSE is on a comparable scale. This is necessary to compare between SPCS and 8n results. Otherwise the data are on different scales so the metrics are not fair to compare.

**Comment by ags3927** [comment]
I would also recommend commenting out the gene thresholding line in the SPCS smoothing notebook. It does not affect the smoothing quality significantly and prevents you having to deal with a reduced gene set for each sample. Because otherwise the smoothing will drop certain genes from each sample and then you will need to find the intersection gene set for the remaining genes and use that in your training, which is a huge pain in the ass.

Comment out these lines:
```
# Select high variance genes
gene.zero.cutoff = 0.7
gene.var.cutoff = 0.0
```
